File: lib/src/models/rivium_trace_config.dart
import '../constants/rivium_trace_constants.dart';
import '../services/rivium_trace_logger.dart';

/// Configuration options for RiviumTrace SDK
class RiviumTraceConfig {
  /// API Key from Rivium Console (rv_live_xxx or rv_test_xxx)
  final String apiKey;

  /// Custom API URL for self-hosted RiviumTrace instances
  /// Defaults to the official RiviumTrace cloud service
  final String apiUrl;

  /// Environment name (e.g., 'production', 'staging', 'development')
  final String environment;

  /// Release version
  final String? release;

  /// Whether to capture uncaught errors automatically
  final bool captureUncaughtErrors;

  /// Whether RiviumTrace is enabled
  final bool enabled;

  /// Enable debug logging
  /// When true, SDK logs are sent to dart:developer and can be viewed in DevTools
  final bool debug;

  /// Request timeout in seconds
  final int timeout;

  /// Maximum number of breadcrumbs to keep (default: 20)
  final int maxBreadcrumbs;

  /// Sample rate for error capture (0.0 to 1.0, default: 1.0)
  /// 1.0 means all errors are captured, 0.5 means 50% of errors
  final double sampleRate;

  /// Enable offline storage for errors when network is unavailable
  final bool enableOfflineStorage;

  /// Custom log handler to receive SDK logs
  ///
  /// Use this to forward logs to your own logging system or display them
  /// in a custom debug console. Works in both debug and release modes.
  ///
  /// Example:
  /// ```dart
  /// RiviumTraceConfig(
  ///   apiKey: 'your-api-key',
  ///   debug: true,
  ///   logHandler: (level, message, error, stackTrace) {
  ///     print('[$level] $message');
  ///   },
  /// )
  /// ```
  final RiviumTraceLogCallback? logHandler;

  const RiviumTraceConfig({
    required this.apiKey,
    String? apiUrl,
    this.environment = 'production',
    this.release,
    this.captureUncaughtErrors = true,
    this.enabled = true,
    this.debug = false,
    this.timeout = 30,
    this.maxBreadcrumbs = 20,
    this.sampleRate = 1.0,
    this.enableOfflineStorage = true,
    this.logHandler,
  }) : apiUrl = apiUrl ?? RiviumTraceConstants.apiUrl;

  /// Create a simple configuration with just an API key
  factory RiviumTraceConfig.simple(String apiKey) {
    return RiviumTraceConfig(apiKey: apiKey);
  }

  /// Create a copy with modified values
  RiviumTraceConfig copyWith({
    String? apiKey,
    String? apiUrl,
    String? environment,
    String? release,
    bool? captureUncaughtErrors,
    bool? enabled,
    bool? debug,
    int? timeout,
    int? maxBreadcrumbs,
    double? sampleRate,
    bool? enableOfflineStorage,
    RiviumTraceLogCallback? logHandler,
  }) {
    return RiviumTraceConfig(
      apiKey: apiKey ?? this.apiKey,
      apiUrl: apiUrl ?? this.apiUrl,
      environment: environment ?? this.environment,
      release: release ?? this.release,
      captureUncaughtErrors:
          captureUncaughtErrors ?? this.captureUncaughtErrors,
      enabled: enabled ?? this.enabled,
      debug: debug ?? this.debug,
      timeout: timeout ?? this.timeout,
      maxBreadcrumbs: maxBreadcrumbs ?? this.maxBreadcrumbs,
      sampleRate: sampleRate ?? this.sampleRate,
      enableOfflineStorage: enableOfflineStorage ?? this.enableOfflineStorage,
      logHandler: logHandler ?? this.logHandler,
    );
  }

  @override
  String toString() {
    return 'RiviumTraceConfig(apiKey: ${apiKey.substring(0, 10)}..., environment: $environment, enabled: $enabled, sampleRate: $sampleRate)';
  }
}

File: android/src/main/kotlin/co/rivium/trace/sdk/RiviumTraceConfig.kt
package co.rivium.trace.sdk

/**
 * Configuration for RiviumTrace SDK
 *
 * @property apiKey The API Key from Rivium Console (rv_live_xxx or rv_test_xxx)
 * @property environment Environment name (e.g., "production", "staging", "development")
 * @property release Release/version string of your application
 * @property debug Enable debug logging
 * @property enabled Enable/disable the SDK entirely
 * @property captureUncaughtExceptions Automatically capture uncaught exceptions
 * @property captureSignalCrashes Capture signal crashes (native crashes via marker detection)
 * @property captureAnr Capture Application Not Responding (ANR) events
 * @property anrTimeoutMs ANR detection timeout in milliseconds (default: 5000ms)
 * @property maxBreadcrumbs Maximum number of breadcrumbs to store (default: 20)
 * @property httpTimeout HTTP request timeout in seconds
 * @property enableOfflineStorage Cache errors when offline for later sending
 * @property sampleRate Sample rate for error capture (0.0 to 1.0)
 * @property apiUrl Base URL of the RiviumTrace server (override for self-hosted instances)
 */
data class RiviumTraceConfig(
    val apiKey: String,
    val environment: String = "production",
    val release: String? = null,
    val debug: Boolean = false,
    val enabled: Boolean = true,
    val captureUncaughtExceptions: Boolean = true,
    val captureSignalCrashes: Boolean = true,
    val captureAnr: Boolean = true,
    val anrTimeoutMs: Long = 5000L,
    val maxBreadcrumbs: Int = 20,
    val httpTimeout: Int = 30,
    val enableOfflineStorage: Boolean = true,
    val sampleRate: Float = 1.0f,
    val apiUrl: String = "https://trace.rivium.co"
) {
    init {
        require(apiKey.isNotBlank()) { "API key cannot be empty" }
        require(apiKey.startsWith("rv_live_") || apiKey.startsWith("rv_test_") || apiKey.startsWith("nl_live_") || apiKey.startsWith("nl_test_")) { "API key must start with rv_live_ or rv_test_" }
        require(maxBreadcrumbs > 0) { "maxBreadcrumbs must be positive" }
        require(httpTimeout > 0) { "httpTimeout must be positive" }
        require(sampleRate in 0.0f..1.0f) { "sampleRate must be between 0.0 and 1.0" }
    }

    /**
     * Builder pattern for creating RiviumTraceConfig
     */
    class Builder(private val apiKey: String) {
        private var environment: String = "production"
        private var release: String? = null
        private var debug: Boolean = false
        private var enabled: Boolean = true
        private var captureUncaughtExceptions: Boolean = true
        private var captureSignalCrashes: Boolean = true
        private var captureAnr: Boolean = true
        private var anrTimeoutMs: Long = 5000L
        private var maxBreadcrumbs: Int = 20
        private var httpTimeout: Int = 30
        private var enableOfflineStorage: Boolean = true
        private var sampleRate: Float = 1.0f
        private var apiUrl: String = "https://trace.rivium.co"

        fun environment(environment: String) = apply { this.environment = environment }
        fun release(release: String?) = apply { this.release = release }
        fun debug(debug: Boolean) = apply { this.debug = debug }
        fun enabled(enabled: Boolean) = apply { this.enabled = enabled }
        fun captureUncaughtExceptions(capture: Boolean) = apply { this.captureUncaughtExceptions = capture }
        fun captureSignalCrashes(capture: Boolean) = apply { this.captureSignalCrashes = capture }
        fun captureAnr(capture: Boolean) = apply { this.captureAnr = capture }
        fun anrTimeoutMs(timeout: Long) = apply { this.anrTimeoutMs = timeout }
        fun maxBreadcrumbs(max: Int) = apply { this.maxBreadcrumbs = max }
        fun httpTimeout(timeout: Int) = apply { this.httpTimeout = timeout }
        fun enableOfflineStorage(enable: Boolean) = apply { this.enableOfflineStorage = enable }
        fun sampleRate(rate: Float) = apply { this.sampleRate = rate }
        fun apiUrl(url: String) = apply { this.apiUrl = url }

        fun build(): RiviumTraceConfig = RiviumTraceConfig(
            apiKey = apiKey,
            environment = environment,
            release = release,
            debug = debug,
            enabled = enabled,
            captureUncaughtExceptions = captureUncaughtExceptions,
            captureSignalCrashes = captureSignalCrashes,
            captureAnr = captureAnr,
            anrTimeoutMs = anrTimeoutMs,
            maxBreadcrumbs = maxBreadcrumbs,
            httpTimeout = httpTimeout,
            enableOfflineStorage = enableOfflineStorage,
            sampleRate = sampleRate,
            apiUrl = apiUrl
        )
    }

    companion object {
        /**
         * Create a simple config with just API key
         */
        fun simple(apiKey: String): RiviumTraceConfig = RiviumTraceConfig(apiKey = apiKey)
    }
}
